src/desktop/usb.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn fd(&self) -> Option<Fd> {
+        self.fd.as_ref().map(|fd| fd.into())
+    }
+}
+
+#[derive(Debug, Deserialize, Type)]
+/// A usb event received part of the `device_event` signal response.
+pub struct UsbEvent(/* action */ String, /* id */ String, UsbDevice);
+
+#[derive(Debug, Deserialize, Type)]
+/// A response received when the `device_event` signal is received.
+pub struct UsbDeviceEvent(OwnedObjectPath, Vec<UsbEvent>);
+
+impl UsbDeviceEvent {
+    /// The session triggered the state change
+    pub fn session_handle(&self) -> ObjectPath<'_> {
+        self.0.as_ref()
+    }
+
+    /// Events received
+    pub fn events(&self) -> &[UsbEvent] {
+        &self.1
+    }
+}
+
+impl UsbEvent {
+    /// The action.
+    pub fn event_action(&self) -> &str {
+        &self.0
+    }
+
+    /// The device id.
+    pub fn event_device_id(&self) -> &str {
+        &self.1
+    }
+
+    /// The [`UsbDevice`] properties.
+    pub fn event_device(&self) -> &UsbDevice {
+        &self.2
+    }
+}
+
+/// This interface provides access to USB devices.
+///
+#[derive(Debug)]
+#[doc(alias = "org.freedesktop.portal.Usb")]
+pub struct UsbProxy<'a>(Proxy<'a>);
+
+impl<'a> UsbProxy<'a> {
+    /// Create a new instance of [`UsbProxy`].
+    pub async fn new() -> Result<UsbProxy<'a>, Error> {
+        let proxy = Proxy::new_desktop("org.freedesktop.portal.Usb").await?;
+        Ok(Self(proxy))
+    }
+
+    /// Create Session
+    ///
+    #[doc(alias = "CreateSession")]
+    pub async fn create_session(&self) -> Result<Session<'a, Self>, Error> {
+        let options = CreateSessionOptions::default();
+        let session: OwnedObjectPath = self.0.call("CreateSession", &(&options)).await?;
+        Session::new(session).await
+    }
+
+    /// Enumerate USB devices.
+    ///
+    #[doc(alias = "EnumerateDevices")]
+    pub async fn enumerate_devices(
+        &self,
+        options: UsbOptions,
+    ) -> Result<Vec<(String, UsbDevice)>, Error> {
+        self.0.call("EnumerateDevices", &(&options)).await
+    }
+
+    /// Acquire devices
+    ///
+    /// The portal will perform the permission request.
+    ///
+    #[doc(alias = "AcquireDevices")]
+    pub async fn acquire_devices(
+        &self,
+        parent_window: &WindowIdentifier,
+        devices: &[(&String, bool)],
+    ) -> Result<Request<()>, Error> {
+        let options = AcquireOptions::default();
+        let acquire_devices: Vec<(String, AcquireDevice)> = devices
+            .iter()
+            .map(|dev| {
+                let device = AcquireDevice { writable: dev.1 };
+                (dev.0.to_string(), device)
+            })
+            .collect();
+        self.0
+            .empty_request(
+                &options.handle_token,
+                "AcquireDevices",
+                &(parent_window, &acquire_devices, &options),
+            )
+            .await
+    }
+
+    /// Call on success of acquire_devices.
+    #[doc(alias = "FinishAcquireDevices")]
```

Comment:
This seems like it can be just called at the end of acquire_device, why it is a separate method?

Reply:
the problem is that there is a dbus limitation one how many fd it can return, so we need to be able to chunk the finish call. So it has to be called manually.

Reply:
We talked about this in the matrix channel, this is a dbus limitation. You should instead call it yourself in AcquireDevices until it stops returning, and return the results or even better make it return a Stream so you can call next on it which would call internally the FinishAcquireDevices.

Reply:
This is now a private method and `AcquireDevices` will return the devices directly.